fms/utils/serialization.py

```diff
@@ -136,12 +194,275 @@ def load_state_dict(
+
+
+def _copy_rowwise_st(
+    module: torch.nn.Module, weights, weight_info_map, rank, world_size, prefix
+):
+    """
+    This function copies the correct shard of the weights for a rowwise-TP'd module
+    according to the rank of the process and the world_size.
+
+    Args
+    ====
+    module: torch.nn.Module
+        Module that has had TP applied
+    weights:
+        safetensors weights object
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    prefix: str
+        Where to find the weight in the weights object
+    """
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = module.weight.shape[1]
+    fms_weight_name = prefix + "weight"
+    if fms_weight_name in weight_info_map:
+        st_name, st_func = weight_info_map[fms_weight_name]
+        tensor_slice = st_func(weights.get_tensor(st_name))
+        tensor = tensor_slice[
+            :,
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+        ]
+        module.weight.copy_(tensor, non_blocking=True)
+    if module.bias is not None:
+        if rank == 0:
+            _copy_if_present_st(module.bias, weights, prefix + "bias", weight_info_map)
+        else:
+            module.bias.zero_()
+
+
+def _copy_embedding_st(
+    module: torch.nn.Module, weights, weight_info_map, rank, world_size, prefix
+):
+    """
+    This function copies the correct shard of the weights for a TP'd embedding module
+    according to the rank of the process and the world_size.
+
+    Args
+    ====
+    module: torch.nn.Module
+        Module that has had TP applied
+    weights:
+        safetensors weights object
+    rank: int
+        Rank of the current process
+    world_size: int
+        Total number of TP processes
+    prefix: str
+        Where to find the weight in the weights object
+    """
+    # Divide the weight matrix along the last dimension.
+    output_size_per_partition = module.weight.shape[1]
+    fms_weight_name = prefix + "weight"
+    if fms_weight_name in weight_info_map:
+        st_name, st_func = weight_info_map[fms_weight_name]
+        tensor_slice = st_func(weights.get_tensor(st_name))
+        tensor = tensor_slice[
+            :,
+            (rank * output_size_per_partition) : (
+                (rank + 1) * output_size_per_partition
+            ),
+        ]
+        module.weight.copy_(tensor, non_blocking=True)
+
+
+def _copy_if_present_st(parameter, model_weights, fms_weight_name, weight_info_map):
+    if fms_weight_name in weight_info_map:
+        st_name, st_func = weight_info_map[fms_weight_name]
+        parameter.copy_(st_func(model_weights.get_tensor(st_name)), non_blocking=True)
+
+
+def _load_safetensors_checkpoint_impl(
+    layer: torch.nn.Module,
+    model_weights,
+    weight_info_map,
+    rank=0,
+    world_size=1,
```

Comment:
I'm a bit confused here... this code seems to assume the model will be tp-sharded. We should support loading HF safetensor weights for other models and other sharding strategies (including unsharded).

It should support, say, loading to only rank0 unsharded, and having `FSDP(model, sync_module_states=True)` handle sharding them from rank0 to other ranks.

Reply:
all these cases are still supported, given that if you aren't doing TP we'll just hit the path under `# If we're on a leaf module and no TP has happened just copy the weights`. I can be more explicit and separate the non-TP path under a whole different function if you feel it's going to be more clear, but it's basically copy-pasting code